src/manifold/src/constructors.cpp

```diff
@@ -317,95 +317,127 @@ Manifold Manifold::Extrude(const CrossSection& crossSection, float height,
-          int lastSlice = (slice == 0 ? nDivisions : slice) - 1;
-          float phi = slice * dPhi;
-          glm::vec2 pos = poly[polyVert];
-          vertPos.push_back({pos.x * cosd(phi), pos.x * sind(phi), pos.y});
-          triVerts.push_back({startVert + slice, startVert + lastSlice,
-                              lastStart + lastSlice});
-          triVerts.push_back(
-              {lastStart + lastSlice, lastStart + slice, startVert + slice});
+
+    for (int polyVert = 0; polyVert < poly.size(); ++polyVert) {
+
+      int startPosIndex = vertPos.size();
+
+      if (!isFullRevolution)
+        startPoses.push_back(startPosIndex);
+
+      int prevStartPosIndex = startPosIndex + (polyVert == 0 ? nRevolveAxisVerts + (nSlices * nPosVerts) - nSlices
+                                                             : poly[polyVert-1].x == 0.0 ? -1 : -nSlices);
+      glm::vec2 currPolyVertex = poly[polyVert];
+      glm::vec2 prevPolyVertex = poly[polyVert == 0 ? poly.size() - 1 : polyVert - 1];
+
+      int nFacesAdded = 0;
+      for (int slice = 0; slice < nSlices; ++slice) {
+
+        float phi = slice * dPhi;
+        if (slice == 0 || currPolyVertex.x > 0) {
+          vertPos.push_back({currPolyVertex.x * cosd(phi), currPolyVertex.x * sind(phi), currPolyVertex.y});
+          nVertices += 1;
         }
-      }
-    } else {  // poly crosses zero
-      int polyVert = start;
-      glm::vec2 pos = poly[polyVert];
-      do {
-        glm::vec2 lastPos = pos;
-        polyVert = (polyVert + 1) % poly.size();
-        pos = poly[polyVert];
-        if (pos.x > 0) {
-          if (lastPos.x <= 0) {
-            float a = pos.x / (pos.x - lastPos.x);
-            vertPos.push_back({0.0f, 0.0f, glm::mix(pos.y, lastPos.y, a)});
-          }
-          int startVert = vertPos.size();
-          for (int slice = 0; slice < nDivisions; ++slice) {
-            int lastSlice = (slice == 0 ? nDivisions : slice) - 1;
-            float phi = slice * dPhi;
-            glm::vec2 pos = poly[polyVert];
-            vertPos.push_back({pos.x * cosd(phi), pos.x * sind(phi), pos.y});
-            if (lastPos.x > 0) {
-              triVerts.push_back({startVert + slice, startVert + lastSlice,
-                                  startVert - nDivisions + lastSlice});
-              triVerts.push_back({startVert - nDivisions + lastSlice,
-                                  startVert - nDivisions + slice,
-                                  startVert + slice});
-            } else {
-              triVerts.push_back(
-                  {startVert - 1, startVert + slice, startVert + lastSlice});
-            }
+
+        if (isFullRevolution || slice > 0) {
+          int lastSlice = (slice == 0 ? nDivisions : slice) - 1;
+          if (currPolyVertex.x > 0.0) {
+            triVerts.push_back({startPosIndex + slice,
+                startPosIndex + lastSlice,
+                // "Reuse" vertex of first slice if it lies on the revolve axis
```

Comment:
Is this the part you were saying messes up the genus? Can you give an example?

Reply:
Kinda. Genus "breaks" when you don't have any special handling of vertices lying on the Y axis. When x=0, `{pos.x * manifold::cosd(phi), pos.x * manifold::sind(phi), pos.y}` yields coincident vertices and degenerate triangles. For example, when a square is offset from the Y axis Revolve will yield 8 triangles and 4 vertices per slice, whereas if one edge is on the Y axis, Revolve will yield 4 triangles and two vertices per slice. In my initial implementation I didn't have any special handling of such cases but surprisingly it still mostly worked. But Genus was off  I guess because Manifold considered the revolved square to be an infinitesimal toroid.

This is the original loop that "works":

```cpp
    for (const auto& poly : polygons) {
        for (std::size_t polyVert = 0; polyVert < poly.size(); ++polyVert) {

            int startVert = vertPos.size();

            if (!isFullRevolution)
                startPoses.push_back(startVert);

            // first and last slice are distinguished if not a full revolution.
            int nSlices = isFullRevolution ? nDivisions : nDivisions + 1;
            int lastStart = startVert + (polyVert == 0 ? nSlices * (poly.size() - 1) : -nSlices);

            for (int slice = 0; slice < nSlices; ++slice) {

                float phi = slice * dPhi;
                glm::vec2 pos = poly[polyVert];
                glm::vec3 p = {pos.x * manifold::cosd(phi), pos.x * manifold::sind(phi), pos.y};
                vertPos.push_back(p);

                int lastSlice = (slice == 0 ? nDivisions : slice) - 1;
                if (isFullRevolution || slice > 0) {
                    triVerts.push_back({startVert + slice, startVert + lastSlice,
                            lastStart + lastSlice});
                    triVerts.push_back(
                        {lastStart + lastSlice, lastStart + slice, startVert + slice});
                }

            }
            if (!isFullRevolution)
                endPoses.push_back(vertPos.size() -1);
        }
    }
```

Reply:
Right, so the problem is without special handling of the center verts you end up with an infinitesimal hole through it like a bead. Definitely important to avoid that, since I don't think our decimator can fix it. If your test case doesn't already check that, it's probably a good idea to add a test for that specifically.